### PR TITLE
Adding Sub task planner for training and inferencing in pi05

### PR DESCRIPTION
**Pull request by akshay18iitg**
## What this does
Adds response prediction for training and inferencing.

## How it was tested
Run training and inferencing on local computer with debug mode
Fine-tuned on oracle nodes. The performance is retained

Examples:
- Adding methods to prepare response for training
- made response field flexible to be represented as subtask, VQA outputs, bounding boxes, etc
- adding autoregressive prediction of response while inferencing
- Optimized using KV cache
- Started using eos as separator instead of /n
- config to on , off response prediction while inferencing
- Updated pi05_integration pytest
- Changed the past_key_values positions

## How to checkout & try? (for the reviewer)
Provide a simple way for the reviewer to try out your changes.

Examples:
```bash
python src/opentau/scripts/train.py --config_path=configs/examples/pi05_training_config.json
```

## Checklist

- [x] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [x] My PR includes policy-related changes.
  - [x] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

### CODE REVIEW

File: src/opentau/policies/pi05/configuration_pi05.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ class PI05Config(PreTrainedConfig):
     # Shorter state and action vectors will be padded
     max_state_dim: int = 32
     max_action_dim: int = 32
+    subtask_prediction: bool = False
```
Comment (shuheng-liu) [inline src/opentau/policies/pi05/configuration_pi05.py]: Let's add this in the docstring as well.

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 2]
```diff
@@ -757,7 +764,14 @@ def prepare_language(self, batch: dict[str, Tensor]) -> tuple[Tensor, Tensor]:
 
         # add state to the prompt
         state = self.prepare_discrete_state(batch)
-        prompt = [f"Task: {task}State: {state}\nActions:" for task, state in zip(tasks, state, strict=False)]
+        if self.config.subtask_prediction:
```
Comment (shuheng-liu) [inline src/opentau/policies/pi05/modeling_pi05.py]: 1. Can we add a comment here explaining why there is a `"\n"` between state and subtask, but no `"\n"` between task and state? We know it's because the `task` string comes with a trailing `"\n"`. Still, without a comment, it looks like a typo. 

2. In the meantime, it's worth mentioning that subtasks and (discrete) actions are mutually exclusive.

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 3]
```diff
@@ -956,6 +1008,20 @@ def embed_prefix(
         num_lang_embs = lang_emb.shape[1]
         att_masks += [0] * num_lang_embs
 
+        if self.config.subtask_prediction and sublang_tokens is not None:
+            sublang_emb = self.paligemma_with_expert.embed_language_tokens(sublang_tokens)
+
+            # Normalize subtask language embeddings
+            sublang_emb_dim = sublang_emb.shape[-1]
+            sublang_emb = sublang_emb * math.sqrt(sublang_emb_dim)
```
Comment (shuheng-liu) [inline src/opentau/policies/pi05/modeling_pi05.py]: Are we confident that we should multiply by sqrt(d) here? I thought we monkey-patched the tokenizer such that it doesn't divide by sqrt(d), hence there is no need to undo that scaling. 

I might be wrong. Please double check.

**Review by shuheng-liu — CHANGES_REQUESTED** [review]
Thank you Akshay for this PR. This is a tricky change and we need to extra cautious. My first round of review only addresses some of the issues I can spot. I might add more in the near future. 

A question: are we removing support for VQA response in pi0.5 and repurposing those tokens for subtask?

File: src/opentau/datasets/lerobot_dataset.py [hunk 4]
```diff
@@ -1506,6 +1506,8 @@ def __getitem__(self, idx) -> dict:
             # change data naming to standard data format
             item = self._to_standard_data_format(item)
 
+            item["subtask"] = item["prompt"]
```
[the remark this answers is not in the document]
Comment (akshay18iitg) [reply src/opentau/datasets/lerobot_dataset.py]: will dynamically load subtask from dataset in future, once we start having datasets for long horizon tasks.
For simple tasks, having subtask same as prompt makes more sense.

Difficult to add load subtask function now, as storage design for subtask in dataset has not been discussed

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 5]
```diff
@@ -1236,7 +1377,7 @@ def denoise_step(
             cross_att_pad_masks=prefix_pad_masks[:, :num_cross_att_tokens],
         )
         # We should skip the response tokens when numbering the position ids for the action expert
-        prefix_offsets = torch.sum(prefix_pad_masks[:, : -self.config.discrete_action_max_length], dim=-1)[
+        prefix_offsets = torch.sum(prefix_pad_masks, dim=-1)[
             :, None
         ]  # action expert position ids start after prefix
```
[the remark this answers is not in the document]
Comment (akshay18iitg) [reply src/opentau/policies/pi05/modeling_pi05.py]: during evaluation, prefix_pad only contains images, language+state and subtask. There are no discrete actions. So, no need to subtract self.config.max_discrete_action

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 6]
```diff
@@ -956,6 +1008,20 @@ def embed_prefix(
...
+            # Normalize subtask language embeddings
+            sublang_emb_dim = sublang_emb.shape[-1]
+            sublang_emb = sublang_emb * math.sqrt(sublang_emb_dim)
+
+            embs.append(sublang_emb)
+            pad_masks.append(sublang_masks)
+
+            # full attention between image, language and subtask inputs
+            num_sublang_embs = sublang_emb.shape[1]
+            att_masks += [1] * num_sublang_embs
```
[the remark this answers is not in the document]
Comment (akshay18iitg) [reply src/opentau/policies/pi05/modeling_pi05.py]: In inference time, subtask will be predicted autoregressively, so need to have autoregressive attention mask while training.

File: src/opentau/policies/pi05/configuration_pi05.py (hunk 1 above)
Comment (akshay18iitg) [reply src/opentau/policies/pi05/configuration_pi05.py]: According to pi0.5 paper, 
The main purpose of these tokens is to predict subtask.
While pre-training, we can use them for subtask prediction as well as VQA response.
While finetunning and inferencing, it is just used for subtask prediction.


Also, our initial release doesn't support VQA response for pi0.5

File: src/opentau/policies/pi05/modeling_pi05.py (hunk 2 above)
Comment (akshay18iitg) [reply src/opentau/policies/pi05/modeling_pi05.py]: done

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 7]
```diff
@@ -757,7 +764,15 @@ def prepare_language(self, batch: dict[str, Tensor]) -> tuple[Tensor, Tensor]:
 
         # add state to the prompt
         state = self.prepare_discrete_state(batch)
-        prompt = [f"Task: {task}State: {state}\nActions:" for task, state in zip(tasks, state, strict=False)]
+        # we use \n as a end of token indictaor. The task always ends with \n , so no need to add it between State and Task. But for response prediction, we need to add it between Task and response.
+        if self.config.response_prediction:
+            prompt = [
+                f"Task: {task}State: {state}\nResponse:" for task, state in zip(tasks, state, strict=False)
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: feat request: let's see if we can replace these newlines with <EOS> tokens

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 8]
```diff
@@ -1141,7 +1224,34 @@ def forward(
...
+        # remove pad tokens
+        response_is_pad = ~response_masks  # convert into format where value for pad is True
+        response_ce_loss = response_ce_loss * ~response_is_pad
+
+        # compute mean
+        response_ce_loss = response_ce_loss.mean()
+
+        return {
+            "MSE": losses,
+            "CE": (ce_loss + response_ce_loss),
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: rename ce_loss to discrete_action_ce_loss

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 9]
```diff
@@ -1179,16 +1289,57 @@ def sample_actions(
...
-            use_cache=self.config.use_cache,
+            use_cache=False,
             fill_kv_cache=True,
         )
 
+        response_tokens = []
+        response_masks = []
+        if self.config.response_prediction:
+            for _ in range(self.config.response_max_length):
+                response_token = prefix_out[:, -1:]
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: make sure BOS token is added before extracting response

File: src/opentau/policies/pi05/paligemma_with_expert.py [hunk 10]
```diff
@@ -416,27 +416,24 @@ def forward(
             query_states = apply_rope(query_states, position_ids)
             key_states = apply_rope(key_states, position_ids)
 
-            if use_cache and past_key_values is None:
+            if past_key_values is None:
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/paligemma_with_expert.py]: I think it's better if we just move this into the fill_kv_cache if statement block

**Review by WilliamYue37 — CHANGES_REQUESTED** [review]

File: src/opentau/policies/pi05/configuration_pi05.py [hunk 11]
```diff
@@ -97,9 +98,12 @@ class PI05Config(PreTrainedConfig):
     # left and right wrist cameras in addition to the top camera.
     empty_cameras: int = 0
 
-    # Tokenizer
+    # Language Tokenizer
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/configuration_pi05.py]: Perhaps this should be updated to prompt_max_length

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 12]
```diff
@@ -753,11 +760,20 @@ def prepare_language(self, batch: dict[str, Tensor]) -> tuple[Tensor, Tensor]:
         tasks = batch["prompt"]
 
         # PaliGemma prompt has to end with a new line
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: delete comment

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 13]
```diff
@@ -772,6 +788,41 @@ def prepare_language(self, batch: dict[str, Tensor]) -> tuple[Tensor, Tensor]:
...
+                - response_masks: Tensor of response language attention masks.
+        """
+
+        if not self.config.predict_response:
+            return None, None
+        device = batch["state"].device
+        responses = batch["response"]
+
+        # PaliGemma response has to end with a new line
+        responses = [response if response.endswith("\n") else f"{response}\n" for response in responses]
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: Please remove

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 14]
```diff
@@ -1207,50 +1208,51 @@ def forward(
...
-            - self.config.discrete_action_max_length
-            - 1 : -self.config.discrete_action_max_length - 1,
+            response_token_start:response_token_end,
         ]
         response_logits = self.paligemma_with_expert.paligemma.lm_head(response_out)
 
         response_logits = response_logits.to(dtype=torch.float32)  # upcast to float32 for loss calculation
         response_logits = rearrange(response_logits, "b s d -> (b s) d")
-        response_labels = rearrange(response_tokens, "b s -> (b s)")
+        response_labels = rearrange(response_tokens[:, 1:], "b s -> (b s)")
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: add comment

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 15]
```diff
@@ -1300,41 +1304,65 @@ def sample_actions(
...
                 response_att_2d_masks = make_att_2d_masks(
-                    prefix_pad_masks[:, -1:],
-                    prefix_att_masks[:, -1:],
+                    response_pad_masks,
+                    response_att_masks,
                     n_cross_att_tokens=num_cross_att_tokens - 1,
                     cross_att_pad_masks=prefix_pad_masks[:, : num_cross_att_tokens - 1],
                 )
-                prefix_position_ids = torch.cumsum(prefix_pad_masks, dim=1)[:, -1:] - 1
+                prefix_offsets += 1
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: can not assume scalar here. This will be a tensor with batched inputs.

File: src/opentau/policies/pi05/modeling_pi05.py [hunk 16]
```diff
@@ -1434,3 +1393,94 @@ def denoise_step(
...
+            # get the last predicted token from the prefix output
+            response_token = prefix_out[:, -1:]
+            response_token = self.paligemma_with_expert.paligemma.lm_head(response_token).argmax(dim=-1)
+
+        pad_token = self.language_tokenizer.pad_token_id
+        # Create pad masks: False if previous token was EOS or PAD
+        if response_tokens.shape[1] > 1:
+            prev_tokens = response_tokens
+            has_eos = (prev_tokens == eos_token).any(dim=1, keepdim=True)
+            has_pad = (prev_tokens == pad_token).any(dim=1, keepdim=True) if pad_token is not None else False
```
Comment (WilliamYue37) [inline src/opentau/policies/pi05/modeling_pi05.py]: get rid of pad_token is None check

File: tests/policies/test_pi05.py [hunk 17]
```diff
@@ -393,9 +457,27 @@ def capture_embed_suffix(*args, **kwargs):
         # --------------------------------- Run the same test but for select_action --------------------------------------
         captured_variables_select_action = {}
 
+        def capture_infer_response(*args, **kwargs):
```
Comment (WilliamYue37) [inline tests/policies/test_pi05.py]: Also check embeddings are the right length

File: tests/policies/test_pi05.py [hunk 18]
```diff
@@ -23,6 +24,37 @@
 class TestPI05Integration:
     """Integration tests for the complete PI05 pipeline."""
 
+    def _check_autoregressive_prefix_pads(self, prefix_pad_masks, response_tokens, prev_prefix_pad_masks):
+        assert prefix_pad_masks.shape[0] == 1
+        assert prefix_pad_masks.shape[1] == prev_prefix_pad_masks.shape[1] + 1
+        assert prefix_pad_masks.dtype == torch.bool
+
+        has_eos = (response_tokens == 1).any(dim=1, keepdim=True)
```
Comment (WilliamYue37) [inline tests/policies/test_pi05.py]: We should make constant variables for these constants

**Review by WilliamYue37 — CHANGES_REQUESTED** [review]

File: docs/source/tutorials/inference.rst [hunk 19]
```diff
@@ -15,3 +15,38 @@ To run inference, run the following command:
...
 
     python lerobot/scripts/inference.py --config_path=outputs/train/pi05/checkpoints/000040/train_config.json
+
+
+Running inference with autoregressive response prediction
+----------------------------------------------------------
+
+To run inference with autoregressive response prediction, you will need to use the ``pi05_inference_config.json`` file.
+Set the predict_response flag to true in the policy config.
+For now, we only support greedy encoding for response prediction.
```
Comment (WilliamYue37) [inline docs/source/tutorials/inference.rst]: encoding -> decoding
